data_structures/raster_source.hpp

```diff
@@ -0,0 +1,185 @@
+
+        boost::filesystem::ifstream stream(boost::filesystem::path{filepath});
+        if (!stream)
+        {
+            throw osrm::exception("Unable to open raster file.");
+        }
+
+        stream.seekg(0, std::ios_base::end);
+        std::string buffer;
+        buffer.resize(stream.tellg());
+
+        stream.seekg(0, std::ios_base::beg);
+
+        BOOST_ASSERT(buffer.size() > 1);
+        stream.read(&buffer[0], buffer.size());
+
+        boost::algorithm::trim(buffer);
+
+        auto itr = buffer.begin();
+        auto end = buffer.end();
+
+        bool r = false;
+        try
+        {
+            r = boost::spirit::qi::parse(
+                itr, end, +boost::spirit::qi::int_ % +boost::spirit::qi::space, _data);
+        }
+        catch (std::exception const &ex)
+        {
+            throw osrm::exception(
+                std::string("Failed to read from raster source with exception: ") + ex.what());
+        }
+
+        if (!r || itr != end)
+        {
+            throw osrm::exception("Failed to parse raster source correctly.");
+        }
+    }
+
+    RasterGrid(const RasterGrid &) = default;
+    RasterGrid &operator=(const RasterGrid &) = default;
+
+    RasterGrid(RasterGrid &&) = default;
+    RasterGrid &operator=(RasterGrid &&) = default;
+
+    std::int32_t operator()(std::size_t x, std::size_t y) { return _data[y * xdim + x]; }
+    const std::int32_t operator()(std::size_t x, std::size_t y) const
+    {
+        return _data[(y-1) * xdim + (x-1)];
+    }
+
+  private:
+    std::vector<std::int32_t> _data;
+    std::size_t xdim, ydim;
+};
+
+/**
+    \brief Stores raster source data in memory and provides lookup functions.
+*/
+class RasterSource
+{
+  private:
+    const float xstep;
+    const float ystep;
+
+    float calcSize(double min, double max, unsigned count) const;
+
+  public:
+    RasterGrid raster_data;
+
+    const int width;
+    const int height;
+    const int xmin;
+    const int xmax;
+    const int ymin;
+    const int ymax;
+
+    RasterDatum getRasterData(const float lon, const float lat) const;
+
+    RasterDatum getRasterInterpolate(const float lon, const float lat) const;
+
+    RasterSource(RasterGrid _raster_data,
+                 std::size_t width,
+                 std::size_t height,
+                 int _xmin,
+                 int _xmax,
+                 int _ymin,
+                 int _ymax);
+};
+
+class SourceContainer
+{
+  public:
+    SourceContainer() = default;
+
+    int loadRasterSource(const std::string &source_path,
+                         double xmin,
+                         double xmax,
+                         double ymin,
+                         double ymax,
+                         unsigned nrows,
+                         unsigned ncols);
+
+    RasterDatum getRasterDataFromSource(unsigned int source_id, int lon, int lat);
+
+    RasterDatum getRasterInterpolateFromSource(unsigned int source_id, int lon, int lat);
```

Comment:
Yeah, that's probably the best we can do for now.

In an optimal world we would have a wrapper type around lon, lat that provides only sensible member functions. After all, with the current design what does it even mean to do `lon / lat` or `lon ^ lat`, that's just nonsense but the compiler allows it since it's fine for their int types.

By the way that's the reason there is a path type in the boost filesystem library, instead of just passing file paths wrapped in strings around, since there are operations on strings that just do not make sense on paths (and the other way round).